Google.Ads.Gax/src/Util/ExceptionUtilities.cs

```diff
@@ -0,0 +1,197 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using Google.Ads.Gax.Lib;
+using Google.Apis.Auth.OAuth2.Responses;
+using Grpc.Core;
+using System;
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Ads.Gax.Util
+{
+    /// <summary>
+    /// Extensions to <see cref="RpcException"/> class to extract OAuth errors.
+    /// </summary>
+    public static class ExceptionUtilities
+    {
+        /// <summary>
+        /// Parses the OAuth exceptions.
+        /// </summary>
+        /// <param name="e">The aggregate exception that wraps the rpc exception.</param>
+        /// <returns>The parsed exception, or <code>null</code> if the exception cannot be parsed.
+        /// </returns>
+        internal static TokenResponseException ParseOAuthException(AggregateException e) =>
+            ParseOAuthException(ExtractRpcException(e));
+
+        /// <summary>
+        /// Parses the OAuth exceptions.
+        /// </summary>
+        /// <param name="rpcException">The rpc exception</param>
+        /// <returns>The parsed exception, or <code>null</code> if the exception cannot be parsed.
+        /// </returns>
+        internal static TokenResponseException ParseOAuthException(RpcException rpcException)
+        {
+            const string EXCEPTION_TYPE_PREFIX =
+                "Google.Apis.Auth.OAuth2.Responses.TokenResponseException: ";
+            const string ERROR_PREFIX = "Error:";
+            const string DESC_PREFIX = "Description:";
+            const string URI_PREFIX = "Uri:";
+
+            if (rpcException == null)
+            {
+                return null;
+            }
+            if (!(rpcException.StatusCode == StatusCode.Internal ||
+                rpcException.StatusCode == StatusCode.Unavailable))
+            {
+                return null;
+            }
+            int prefixIndex = rpcException.Message.IndexOf(EXCEPTION_TYPE_PREFIX);
+            if (prefixIndex == -1)
+            {
+                return null;
+            }
+            string messageToProcess = rpcException.Message.Substring(
+                prefixIndex + EXCEPTION_TYPE_PREFIX.Length);
+            string[] messageParts = messageToProcess.Split('\n');
+            if (messageParts.Length >= 0)
+            {
+                string lineToParse = messageParts[0];
+                string[] subParts = lineToParse.Split(',').Select(s => s.Trim()).ToArray();
+                // TokenErrorResponse is serialized as
+                // Error:\"{Error}\", Description:\"{ErrorDescription}\", Uri:\"{ErrorUri}\"
+                string error = "";
+                string description = "";
+                string uri = "";
+
+                foreach (string subPart in subParts)
+                {
+                    if (subPart.StartsWith(ERROR_PREFIX))
```

Comment:
Could you avoid all these `Trim('"')` calls if you trimmed on line 72?

Reply:
Nope, see line 74 where I've shown an example of how the string actually looks like.